Smooth Movement 5.4: Rock Bottom (GOLD) (MANY HAVE BEEN ASKING) (#94487)
tgstation/tgstation: .github/guides/SMOOTHNESS.md

## Guide to Smooth Visuals

I'd like to begin this document with two disclaimers:
- This is my view of the game/engine at this precise moment in time. Whenever you write something down it will always rot.
- This is far more information about how to make BYOND behave then you likely need. I'm writing it down because it's important that it's documented, because otherwise our solutions would be absolutely incomprehensible.

### What does smooth mean?

> The game server fires at a set tickrate (20 tps or 0.5ds tick_lag on mainline /tg/). Clients are sent things like map data, animations, etc once each tick, but have their own independant tick rate (100 tps by default, this should be divisable by the server's tick rate). Each client tick animations will continue on along some gradient between the start and end values, etc etc.

Reviewer: ```suggestion
The game server fires at a set tickrate, which is 20 ticks per second (TPS) or 0.5 deciseconds (DS) tick_lag (between ticks) on mainline /tg/. Clients are sent visual packets full of things like map data and animation once each tick but have their own independent tick rate (default being 100 TPS, but any divisible value shall work). Each client tick animations will continue on along some gradient between the start and end values.
```

Author: yeet

> Because the client and server are not well, the same thing, they have their own opinions about how far along in an animation the client is in any given frame. Our goal here is to minimize this difference in opinion, to keep the two as synced together as possible.

Author: sure

Achieving smoothness has other benefits, this idea of client and server desyncs applies to stuff like icon state animations and sound lengths just the same.

### What causes this difference in opinion?

> In order to talk about this properly you need to have a basic understanding of the game tick. See this [writeup on tick order](TICK_ORDER.md) for a baseline.

As discussed before the client and server fire at different tick rates. What this means mechanically is the server will try to send out a packet of information to the client every tick_lag ds, and the client will anticipate a packet of information from the server every tick_lag ds. If the client recieves its packet on regular intervals, and is able to render at its specified framerate, then any animations or effects will happen smoothly, assuming we don't literally fuck up the timing on the server.

There's a bunch of different things that can happen which can fuck up this cycle.

#### 1. Network Lag

A series of tubes separates the client from the server, and if the packet of information containing appearance updates gets slowed down in transit then we'll see that as jitter on the client.

> There isn't a ton we can actually do about this, because sometimes the internet fucks up. We kind of just have to assume the network connection will be stable.

Reviewer: ```suggestion
There isn't a ton we can actually do about this, because sometimes the internet fucks up. We operate under the assumption the network connection will be stable.
```

Author: Seems more clunky to me, is this understandable as is?

#### 2. Clientside Overrun

When set to make a new frame X times a second, the client is expected to make a new frame X times a second. This number is built into the gradients used for any animations happening clientside. If it's not able to meet quota, the user will see delays in action and other sorts of jitter.

> This usually happens because the user's computer just doesn't have a powerful enough GPU (or CPU in some cases) to render a frame in the time required. This is something we can work on, but it's quite complicated since we lack effective profiling tools for clientside issues. You can get a decent feeling for your own machine by looking at the task manager (as a general rule: more stuff on screen/more filters being applied to that stuff -> more clientside lag. There's some flaws here on byond's end which lummox is working on but there's only so much you can do).

Author: Can do

#### 4. Server Underrun

I've been lying to you a little bit so far. We don't actually need the server and client tick rate to sync up perfectly. What we actually need is the time between map sends to sync up with the client's framerate (IE, the time between mapsends needs to be consistent). This is because appearance info is sent by the server as soon as map sending finishes, the time after when verbs run is something closer to "dead" time (verbs don't even show up on byond's cpu tracking tools, although that's for different reasons).

This means that if say, it takes 70% of the tick to get to sending maps on tick 1, 20% on tick 2, and 90% on tick 3, then clients will see that as jitter "forward" (50% of a tick separates 1 and 2) and then a long delay (170% of a tick separates 2 and 3).

> There are ways for byond to fix this jumping "foward" (it could just hold the frame for a little while, though this would be weird with afformentioned network problems), but fixing the delay is harder and would in theory require some form of clientside prediction.

Reviewer: ```suggestion
There are ways for byond to fix this jumping "forward" (it could just hold the frame for a little while, though this would be weird with aforementioned network problems), but fixing the delay is harder and would in theory require some form of client-side prediction.
```

This is because verbs are executed when they arrive at the server. They're literally just requests from the clients to "do something" remember. They're not allowed to run in map sending or while sleeping procs are going, but afterwards it's basically free game for them.

This means in order to keep them from causing server overrun, we need to keep track of when they're trying to run and queue them up if it's too late to finish on time.

> Verbs can also somtimes run AFTER a tick is over, even if no verbs were holding up the tick before. I frankly have no idea why this happens.

#### Types of "failure"

There's two types of ways to fuck up a transition (sound, animate() call, glide, etc). You can start too early, or start too late.

It's quite rare for "undershooting" to be a problem. It only really shows up if the client gets packets too quickly, which would either happen due to server underrun, or because of network hell leading to the client getting a bunch of packets at once.

> Overshooting on the other hand is quite common. We have a very hard time avoiding server overrun because well, code can functionally do whatever it wants. The MC does its best to avoid this but if a subsystem/sleeping proc/verb just decides it's not gonna play ball then we're gonna bite it.

Reviewer: ```suggestion
Overshooting on the other hand is quite common. We have a very hard time avoiding server overrun because code can functionally do whatever it wants. The MC does its best to avoid this but if a subsystem/sleeping proc/verb just decides it's not going to cooperate then we're going to have to eat the cost.
```
